### DISCUSSION TITLE
Supported boards/hardware/machines: riscv64

### DISCUSSION THREAD

**Discussion by eshattow** [home-assistant/architecture#1107] [General] [answered]
Hi all, I tried to install Home Assistant and was surprised this was not already underway for RISC-V. It may be time to allow riscv64 to join the Home Assistant supported machine architectures listing?

Package coverage is very good in upstream Debian GNU/Linux and debian-installer. Hardware riscv64 boards are widely available and comparable in cost to the existing supported Raspberry Pi boards.

An example: I have the StarFive JH7110 CPU installed in Home Assistant Yellow PoE hardware as a CM4-compatible module (Milk-V Mars CM Lite). This same CPU is being used by Framework with their RISC-V developer laptop mainboard announcement, was announced as supported by an official Ubuntu OS image, and has widest distribution from multiple global suppliers in the form of a reference board from StarFive as the VisionFive2, and also Pine64 Star64. There are many other interesting RISC-V CPUs and I mention in particular the JH7110 because of availability as a CM4 compatible module which I know firsthand to be working in official Home Assistant Yellow PoE product with upstream Linux and upstream U-Boot as firmware.

I was referred here from https://github.com/home-assistant/os-agent/issues/191

**Comment by agners** [comment]
Bring-up of a new architecture should start with support in Home Assistant Core. For HAOS and Supervised installations we build on the container version of HA Core, so that is where it has to start. The HA Core containers are built using Alpine, for a while this was a road block. With Alpine 3.20 that road block is finally gone, RISC-V is officially supported :tada: .

The next obstacle is to build all the Python wheels HA Core depends on. We have our own wheels builder (at https://github.com/home-assistant/wheels) which builds wheels for Alpine (musllinux). This would need to be extended to support the `riscv64` architecture. As it seems, currently PyPI (more specifically manylinux images, see https://github.com/pypa/manylinux/discussions/1426) and the tools around PyPI like auditwheel (https://github.com/pypa/auditwheel/issues/408) do not support `riscv64` officially yet. Once `riscv64` is enabled, it will be interesting to see if all the Python wheels build. Quite some have native code which needs to be compiled. Simple native code probably just works, but at times a new architecture still requires changes to make it actually work. YMMV.

The lack of really good board options is a bit of a blocker. Granted the StarFive JH7110 is somewhat viable for this, but is really on the lower end in terms of performance, from what I can tell.

That said, if you'd like to start investigate: Adding `riscv64` support in a fork of the above wheels repo is probably a good way. Once you have a wheels builder, you can run it and see if (all) the wheels are building.

**Reply by eshattow** [reply]
edit 2025-01-04: auditwheel 6.2.0 (released 2025-01-04) adds musllinux RISC-V support. ~I have submitted a pull request home-assistant/wheels/pull/822 bump to auditwheel version 6.2.0 in home-assistant/wheels~
edit 2025-01-14: home-assistant/wheels now has auditwheel version 6.2.0

Is there more required for extending support? The docker builds fine. I'm not sure what to do with it though, it seems depending on GitHub web services which I don't know anything about.

**Reply by eshattow** [reply]
hass-docker-base alpine builds with a patch to `go build` tempio from source because your repo (of course) does not host riscv64 executable, and no problems for building python 3.13 image. Yes, the hass base image fails for no matching distribution of Brotli==1.1.0 for the reason you said it is using the HomeAssistant wheels repository. I will try to figure out how to do this next.

FYI a uv pip install of homeassistant package_constraints.txt in alpine reveals some observations:
1. There are version pinning 1000000000 which I filter out because it confuses uv
2. Software netifaces is a dependency of aiodiscover and has some warnings as errors, so must change CFLAGS to accept these problems.
3. Software av has warnings as errors, so must change CFLAGS to accept these problems.
4. 5 hours total time on JH7110 for manual alpine docker instance and to uv pip install homeassistant, python-matter-server, and then all pip packages from package_constraints.txt
5. alpine:3.21 or alpine:edge as of this writing is python3.12 and not any python3.13 is available, so that direct tactic is of limited use for latest version of HomeAssistant.  However it is at least possible to install homeassistant this way in an alpinelinux instance.

Full commands for hands-on alpine direct experiment:
```
docker run -it alpine:3.21
apk add --no-cache shadow bash uv python3 \
 gcc python3-dev musl-dev libffi-dev \
 rust cargo openssl-dev zlib-dev libjpeg-turbo-dev \
 ffmpeg-dev g++ \
 cmake ninja make autoconf automake

useradd -rm -s /bin/bash homeassistant && \
install -d -o homeassistant -g homeassistant /srv/homeassistant

su -l homeassistant
uv venv /srv/homeassistant
source /srv/homeassistant/bin/activate
uv pip install homeassistant  # 1h42m47s duration on JH7110
uv pip install --quiet python-matter-server --upgrade --constraint /srv/homeassistant/lib/python3.12/site-packages/homeassistant/package_constraints.txt

CFLAGS="-Wno-int-conversion" uv pip install netifaces==0.11.0  # 5s duration on JH7110, dependency of aiodiscover==2.1.0
CFLAGS="-Wno-deprecated-declarations -Wno-discarded-qualifiers -Wno-incompatible-pointer-types" uv pip install av==13.1.0  # 4m31s duration on JH7110

# filter out version pinning:
sed '/1000000000/d' /srv/homeassistant/lib/python3.12/site-packages/homeassistant/package_constraints.txt | uv pip install -r -  # 3h24m44s duration on JH7110, needs: cmake ninja make autoconf automake
```

**Reply by eshattow** [reply]
```
git clone https://github.com/home-assistant/docker-base.git ha-docker-base.git
curl https://freeshell.de/e/riscv64/0001-Update-alpine-Dockerfile-to-build-tempio.patch | git -C ha-docker-base.git am
docker build -t ha-base-alpine --build-arg BUILD_FROM=riscv64/alpine:3.21 --build-arg BUILD_ARCH=riscv64 --build-arg BASHIO_VERSION=0.16.2 --build-arg TEMPIO_VERSION=2024.11.2 --build-arg S6_OVERLAY_VERSION=3.1.6.2 --build-arg JEMALLOC_VERSION=5.3.0 ha-docker-base.git/alpine
docker build -t ha-base-python --build-arg BUILD_FROM=ha-base-alpine --build-arg PYTHON_VERSION=3.13.1 --build-arg PIP_VERSION=24.3.1 --build-arg GPG_KEY=7169605F62C751356D054A26A821E680E5FA6305 ha-docker-base.git/python/3.13
```
Successfully tagged ha-base-alpine:latest 
real    7m40.329s
user    0m0.252s
sys     0m0.182s

Successfully tagged ha-base-python:latest 
real    289m35.252s
user    0m2.935s
sys     0m1.085s

```
git clone https://github.com/home-assistant/docker.git ha-docker.git
curl https://freeshell.de/e/riscv64/0001-Build-requirements.txt-dependencies.patch | git -C ha-docker.git am
docker build -t ha-docker --build-arg BUILD_FROM=ha-base-python --build-arg BUILD_ARCH=riscv64 --build-arg SSOCR_VERSION=2.23.1 --build-arg LIBCEC_VERSION=6.0.2 --build-arg TELLDUS_COMMIT=2598bbed16ffd701f2a07c99582f057a3decbaf3 --build-arg PICOTTS_HASH=e3ba46009ee868911fa0b53db672a55f9cc13b1c ha-docker.git
```
Successfully tagged ha-docker:latest 
real    15m39.739s
user    0m0.270s
sys     0m0.322s

```
git clone https://github.com/home-assistant/wheels.git ha-wheels.git
docker build -t ha-wheels --build-arg BUILD_FROM=ha-base-python --build-arg BUILD_ARCH=riscv64 --build-arg CPYTHON_ABI=cp313 ha-wheels.git
```
Successfully tagged ha-wheels:latest 
real    28m54.994s
user    0m0.201s
sys     0m0.123s

```
git clone https://github.com/home-assistant/core.git ha-core.git
curl https://freeshell.de/e/riscv64/0001-Modification-to-allow-riscv64-architecture-build.patch | git -C ha-core.git am
docker build -t ha-core --build-arg BUILD_FROM=ha-docker --build-arg BUILD_ARCH=riscv64 ha-core.git
```
Successfully tagged ha-core:latest 
real    432m58.306s
user    0m4.735s
sys     0m3.215s

 $ docker image ls

REPOSITORY       TAG       IMAGE ID       CREATED         SIZE
ha-core          latest    50132177be58   2 minutes ago   2.21GB
ha-wheels        latest    4e6a18038898   7 hours ago     947MB
ha-docker        latest    a8f7a30483fd   8 hours ago     252MB
ha-base-python   latest    649c90a4dd5e   11 hours ago    119MB
ha-base-alpine   latest    10eb46226f50   16 hours ago    43.3MB
riscv64/alpine   3.21      01bb16e7269c   2 days ago      6.57MB

```
# alternate without using (missing?) init
mkdir $HOME/home-assistant-config
docker run -d --name="hass-core" -e "TZ=America/Los_Angeles" -v $HOME/home-assistant-config:/config --net=host ha-core hass
```

if I try with `--init` then there is a complaint about missing `cataonit` but anyways aside from having no idea how to use the wheels builder, all the wheels built anyways for home-assistant/core

@agners Yes and hass appears to work when accessed via web browser. Matter server complains about something at runtime but again I'm not sure if init is needed or what I am supposed to do, the point of this was what you said about building wheels though, can you provide additional guidance what I should be doing more?

**Comment by olof-nord** [comment]
Just for the sake of documentation and for future reference - PRs which add RISC-V support to Home Assitant is currently not welcomed.
For example here: https://github.com/home-assistant/tempio/pull/75

What I understand is that once an architectural proposal has been approved, changes such as that one are welcome.

**Comment by eshattow** [comment]
@frenck How should we proceed then?

**Reply by frenck** [reply]
As we communicated before: On our end there is no proceeding happening until there is an architectural approval.

**Comment by eshattow** [comment]
PyPI now accepts risc-v uploads

**Comment by eshattow** [comment]
Given that riscv64 is supported in Alpine Linux 3.20 and support for the previous release 3.19 is end-of-life soon 2025-11-01; also ubuntu and debian OS have stable releases officially with riscv64 support - Let's try to get Home Assistant ready for riscv64 ADR then?

1.) homeassistant/tempio:
  i.) riscv64:
      - .github/workflows/build.yml: copy-paste for additional architecture

2.) homeassistant/cosign:
  i.) <del>submitted https://github.com/home-assistant/cosign/pull/4 update from alpine 3.18 to alpine 3.22</del> (merged)
  ii.)  riscv64:
        - .github/workflows/build.yml: trivial copy-paste for additional architecture

3.) homeassistant/builder:  depends on homeassistant/base published ghcr.io images (cyclical dependency)
  i.  <del>submitted https://github.com/home-assistant/builder/pull/248 update from alpine 3.20 to alpine 3.22</del> (merged)
  ii.)  riscv64:
        - Dockerfile: copy-paste for additional architecture
        - builder.sh: add long option help description, long option handling, append BUILD_LIST entry  
        - build.yaml: copy-paste for additional architecture
        - README.md: add long option help description and example

4.) homeassistant/docker-base: depends on homeassistant/tempio homeassistant/cosign homeassistant/builder
  i.) <del>submitted https://github.com/home-assistant/docker-base/pull/310 fix typo Ubuntu version 20.4 => 20.04</del> instead favor <del>https://github.com/home-assistant/docker-base/pull/295</del> (merged) for this since 20.04 is end-of-life a few months ago anyhow.
  ii.)  cyclical dependency with homeassistant/builder

What's next, do we drop alpine 3.19 and ADR riscv64 as a supported platform then in a few months?

I got stuck at this point since things are self-referential to the homeassistant github instance. I've just started to figure out GitHub actions what that is and so I'll try to fork and hack it up for proof-of-concept building on github infrastructure.

**Comment by eshattow** [comment]
How do I bootstrap the cyclical dependency between homeassistant/docker-base and homeassistant/builder ?

**Reply by agners** [reply]
Hm, from history it seems that the builder existed first. I guess the builder was based on something else at first, and only later based on docker-base.

You can go two ways about it: Manually create a builder first, then push that, then add the docker-base. Or manually create a Docker base image, then push that before creating the first builder.

That said, we likely will rework the whole container setup anyways. Probably some time after we dropped the now deprecated architectures. After that it should hopefully be easier to add RISC-V.

**Reply by eshattow** [reply]
Can we change home-assistant/docker-base workflow to build and publish for ${{ github.repository }} instead of hard-coded to 'home-assistant/docker-base' ? The most friction of making any improvements is that when I clone the home-assistant/docker-base repository it fails to publish (on some other github organization) because it is hard-coded, so I just gave up trying. Being able to clone and publish 'docker-base' would give a basis for subsequently clone and build 'builder'.

I think we could overlook the cyclical dependency of 'builder' if 'docker-base' could clone and publish to a different github organization, and then scaffold in some modifications to a cloned 'builder' repository to then use the cloned 'docker-base'... etc.

I attempted to this change myself but ran into trouble where I don't understand how we can use github var and env in labels i.e. `alpine/build.yaml` it seems out of scope to be using these definitions as a value of key labels: org.opencontainers.image.source i.e.

**Reply by eshattow** [reply]
okay I see it now those configurations like `alpine/build.yaml` are not in the GitHub Actions workflow,  they are instead as configuration for home-assistant/builder and support for a substitution similar to `{arch}` would need to be from builder.sh there.

**Comment by jobsidi** [comment]
Seems there are some progress
https://discuss.python.org/t/packaging-support-for-riscv64/58475/13
quay.io/pypa/manylinux_2_39_riscv64

**Comment by eshattow** [comment]
I'm tracking progress in https://github.com/home-assistant/builder/issues/260 to make it possible to fork (clone) https://github.com/home-assistant/builder elsewhere on GitHub in a different namespace and have the actions bootstrap from home-assistant namespace instead of being 100% hard-coded to home-assistant namespace. This should resolve some of the chicken-and-egg problem.

Update Mar 2026:  There is now a move to use multi-arch images and home-assistant/builder is deprecated, so will try this soon to assess how it is now to build for riscv64

**Comment by eshattow** [comment]
The recursive hurdle of getting docker-base and builder image to build is resolved by the recent transition of builder to GitHub Actions. In this way it is not necessary to build or publish builder repo for testing a new architecture it is only needed to commit changes and reference that commit tag in docker-core. Issues to track going forward:

- [ ] home-assistant/tempio#75 (pending ADR)
- [ ] sigstore/cosign-installer#228
- [x] riseproject-dev/riscv-runner-images#38
- [x] home-assistant/builder#303
- [x] home-assistant/docker-base#360
- [x] home-assistant/docker-base/pull/361
- [x] home-assistant/wheels#1019
- [x] home-assistant/wheels#1022
- [x] home-assistant/wheels#1023
- [ ] home-assistant/wheels#1024
- [x] home-assistant/wheels#1035
- [x] home-assistant/wheels#1037
- [ ] <del>home-assistant/docker#405</del>
- [ ] <del>home-assistant/docker#406</del>


Other related dependencies of Home Assistant core to track:

- [ ] AlexxIT/go2rtc#2236

Upstream wheels tracking issues for build and publish on riscv64
- [ ] python-pillow/Pillow#9462
- [ ] grpc/grpc#41591
- [ ] orjson (e-mailed author, there is no issue tracker and forge page closed to pull requests)
- [ ] pynacl (depends: pyca/infra#749)
- [ ] pyca/cryptography#14460 (depends: pyca/infra#749)
pycares
- [ ] python-zeroconf/python-zeroconf#1665
- [ ] Bluetooth-Devices/dbus-fast#620

Using the [RISE RISC-V runners (personal)](https://github.com/settings/installations/119635444) GitHub App then adding riscv64 where appropriate and swapping out the runner label to use RISE RISC-V runners, for builder and docker-base, and substitute for reference to the modified builder repo where appropriate throughout docker-base, I have a successful build of docker-base and docker-wheels.

Build times of docker-base GHA for single riscv64 architecture:
  8m 6s test riscv64 build base images (fails initially due to missing published base docker image but is successful on re-run after release build completes)
  0m 14s test riscv64 build release drafter
  4h 3m 57s riscv64 build base images

Build times of wheels GHA for single riscv64 architecture:
  27m 46s test riscv64 build publish
  12m 2s test riscv64 build CI
  0m 8s test riscv64 build release drafter

wheels tests build by bumping the alpine instance to alpine 3.23 because of the requirement for rustc >=1.89 in orjson

Wheels can fully build and making progress on docker <del>, there are just some quirks about the wheels extra packages repository (https) that is difficult to reason about where I should change this URL as it is hard-coded in several places.</del>  I've got most of the confusion now resolved around wheels index URL usage, pending PR's to home-assistant/wheels.

**Comment by eshattow** [comment]
wheels are built, and homeassistant-base riscv64 is built. I've only built so far riscv64 arch for python 3.14 and alpine 3.23 which deprecates some CMake compatibility mode, so a few more patches are due before getting riscv64 to build together with the other architectures.

**Reply by olof-nord** [reply]
Thanks for all your work on this!

**Reply by eshattow** [reply]
Linking to gounthar/riscv64-python-wheels#15 which is a meta-issue not related to Home Assistant it is for a more general problem-solving of missing riscv64 compiled wheels at PyPI

**Comment by eshattow** [comment]
There's a bit of struggle getting there but I've achieved first build of riscv64-homeassistant, and I'm using that now on my Home Assistant Yellow baseboard with installed Milk-V Mars CM.

Notable challenges to building core include:

- translations depend on an external provider that requires payment
- workflow depends on arch-specific go2rtc docker image
- go2rtc adjacent to PyTurboJPEG wheel depends on libjpeg-turbo-dev version 2.x which is its own complicated thing because Debian and Ubuntu distros had a divergent strategy the past 15 years for compatibility with the old JPEG reference implementation API so it is not consistent between alpine and ubuntu and debian for running some of the scripts that update files during the workflow
- some noticeably time-consuming wheels builds are invoked because those are just assumed to be available from PyPI  and not included in the wheels builder
- warnings as errors when building unmaintained netifaces wheel
- workflow guards against building from forked repository

I'm now backtracking to figure out what the minimum viable changes needed are, and to try to get OS image to build.

**Comment by edenhaus** [comment, marked as answer]
Thanks for raising this and for your enthusiasm around RISC-V! 🙂

We've discussed this internally and for the time being we don't plan to add official support for riscv64 in Home Assistant. The main reason is the current lack of widely available, popular consumer boards — our supported architecture list is driven by real hardware that a meaningful portion of our user base actually runs HA on, and RISC-V isn't quite there yet in terms of mainstream adoption. Adding support also means committing to maintaining it long-term, and we're only willing to take that on when there are enough boards and real-world usage to justify it.

That said, this being open source, nothing stops the community from continuing to work on RISC-V support outside of core. We'd just ask that changes driven by an unsupported architecture aren't pushed into core in a way that creates maintenance burden for the rest of the project. Concretely, this means we won't revert changes or hold up fixes because of issues that only affect unsupported architectures.

If community members reach out to upstream dependencies to request RISC-V support, please make it clear that `riscv64` is not an officially accepted architecture in Home Assistant and that you are not speaking on behalf of the HA team. We want to avoid creating false expectations with upstream projects or giving the impression that this is an effort backed by the project.

If RISC-V gains more traction and popular boards become widely available, we'd genuinely love to revisit this — please feel free to open a new architecture discussion at that point and we'll be happy to reconsider. 🚀

**Reply by eshattow** [reply]
There are challenges yet to be resolved, it is premature to decide about this.

**Reply by edenhaus** [reply]
@eshattow Please leave the above comment selected an answer as we decided on this proposal for now. As written above, we can reconsider it in the future and there please open a new one.

> There are challenges yet to be resolved, it is premature to decide about this.

You should not open a discussion if it's not ready to discuss. We decided what is available currently and we can change our mind in the future if more RISCV boards are available. I know you don't like our conclusion but we need to maintain core and currently we don't want to add support for RISCV
